pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/apply_trait_test.go

```diff
@@ -280,11 +270,57 @@ spec:
 			By("Check added field")
 			v, _, _ = unstructured.NestedString(traitObj.UnstructuredContent(), "spec", "added")
 			Expect(v).Should(Equal("bar"))
-			// if use patching trait, this field will not be removed
 			By("Check removed field")
 			_, found, _ = unstructured.NestedString(traitObj.UnstructuredContent(), "spec", "removed")
 			Expect(found).Should(Equal(false))
 		})
 	})
 
+	// others means anything except AppConfig controller
+	// e.g. trait controllers
+	When("trait instance is changed by others", func() {
+		// if others make changes on the fields managed by AppConfig controller
+		// these changes will reverted but not retained.
+		It("should retain changes by others", func() {
+			By("Get the trait newly created")
+			Eventually(func() string {
+				if err := k8sClient.Get(ctx, appConfigKey, &appConfig); err != nil {
+					return ""
+				}
+				if appConfig.Status.Workloads == nil {
+					reconcileRetry(reconciler, req)
+					return ""
+				}
+				return appConfig.Status.Workloads[0].Traits[0].Reference.Name
+			}, 5*time.Second, time.Second).ShouldNot(BeEmpty())
+			traitName := appConfig.Status.Workloads[0].Traits[0].Reference.Name
+			var traitObj unstructured.Unstructured
+			traitObj.SetAPIVersion("example.com/v1")
+			traitObj.SetKind("Bar")
+			Expect(k8sClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: traitName}, &traitObj)).Should(Succeed())
+
+			By("Others change the trait: add a field")
+			unstructured.SetNestedField(traitObj.Object, "bar", "spec", "added")
+			Expect(k8sClient.Patch(ctx, &traitObj, client.Merge)).Should(Succeed())
+
+			By("Check the change works")
+			var changedTrait unstructured.Unstructured
+			changedTrait.SetAPIVersion("example.com/v1")
+			changedTrait.SetKind("Bar")
+			Expect(k8sClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: traitName}, &changedTrait)).Should(Succeed())
+			v, _, _ := unstructured.NestedString(changedTrait.UnstructuredContent(), "spec", "added")
+			Expect(v).Should(Equal("bar"))
+
+			By("Reconcile")
+			reconcileRetry(reconciler, req)
+
+			By("Check the change is retained")
+			changedTrait = unstructured.Unstructured{}
+			changedTrait.SetAPIVersion("example.com/v1")
+			changedTrait.SetKind("Bar")
+			Expect(k8sClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: traitName}, &changedTrait)).Should(Succeed())
+			v, _, _ = unstructured.NestedString(changedTrait.UnstructuredContent(), "spec", "added")
+			Expect(v).Should(Equal("bar"))
```

Comment:
It makes sure that normal reconcile won't change the added spec.

Can you add one more action below:

1. update the trait in AppConfig, add a new field, and delete an old field added by AppConfig before.
2.  reconcile it
3. check the third-party added field is still there, and check the new field added, the old field deleted as expected.